assets/js/modules/analytics/datastore/report.js

```diff
@@ -183,6 +184,78 @@ const baseSelectors = {
 
 		return reports[ stringifyObject( options ) ];
 	},
+
+	/**
+	 * Gets a Page title to URL map for the given options.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state             Data store's state.
+	 * @param {Object} report            A report from getReport selector containing pagePaths.
+	 * @param {Object} options           Options for generating the report.
+	 * @param {string} options.startDate Required, start date to query report data for as YYYY-mm-dd.
+	 * @param {string} options.endDate   Required, end date to query report data for as YYYY-mm-dd.
+	 * @return {(Object|undefined)} A map with url as the key and page title as the value. `undefined` if not loaded.
+	 */
+	getPageTitles: createRegistrySelector(
+		( select ) => ( state, report, { startDate, endDate } = {} ) => {
+			if ( ! Array.isArray( report ) ) {
+				return;
+			}
+			const pagePaths = []; // Array of pagePaths.
+			const REQUEST_MULTIPLIER = 5;
+			/*
+			 * Iterate the report, finding which dimension contains the
+			 * ga:pagePath metric which we add to the array of pagePaths.
+			 */
+			( report || [] ).forEach( ( { columnHeader, data } ) => {
+				if (
+					Array.isArray( columnHeader?.dimensions ) &&
+					Array.isArray( data?.rows ) &&
+					columnHeader.dimensions.includes( 'ga:pagePath' )
+				) {
+					const pagePathIndex = columnHeader.dimensions.indexOf(
+						'ga:pagePath'
+					);
+					( data?.rows || [] ).forEach( ( { dimensions } ) => {
+						if (
+							! pagePaths.includes( dimensions[ pagePathIndex ] )
+						) {
+							pagePaths.push( dimensions[ pagePathIndex ] );
+						}
+					} );
+				}
+			} );
```

Comment:
If `pagePaths` is empty, let's return early here to avoid a useless request. Here it should probably return an empty object though to prevent an endless loading state.

Reply:
https://github.com/google/site-kit-wp/pull/3958/commits/95a9b65d0330d3479871a83f2edb031e0ecb820d